File: backend/app/models/offre.py
from sqlalchemy import Column, Integer, String, Text, Boolean, DateTime, Enum, ForeignKey
from sqlalchemy.orm import relationship
from datetime import datetime
import enum
from ..database import Base


class OffreStatus(str, enum.Enum):
    PENDING = "pending"
    VALIDATED = "validated"
    REJECTED = "rejected"


class Offre(Base):
    __tablename__ = "offres"
    
    id = Column(Integer, primary_key=True, index=True)
    titre = Column(String(255), nullable=False)
    description = Column(Text, nullable=False)
    type_formation = Column(String(100))  # Licence, Master, Doctorat, etc.
    duree = Column(String(50))  # "2 ans", "3 ans", etc.
    conditions = Column(Text)
    status = Column(Enum(OffreStatus), default=OffreStatus.VALIDATED, nullable=False)  # Admin creates validated offers by default
    admin_id = Column(Integer, ForeignKey("users.id"), nullable=False)
    created_at = Column(DateTime, default=datetime.utcnow)
    updated_at = Column(DateTime, default=datetime.utcnow, onupdate=datetime.utcnow)
    
    # Relations
    admin = relationship("User", back_populates="offres")
    candidatures = relationship("Candidature", back_populates="offre", cascade="all, delete-orphan")

File: frontend/lib/models/offre.dart
class Offre {
  final int id;
  final String titre;
  final String description;
  final String? typeFormation;
  final String? duree;
  final String? conditions;
  final String status;
  final int adminId;
  final DateTime createdAt;
  final DateTime updatedAt;

  Offre({
    required this.id,
    required this.titre,
    required this.description,
    this.typeFormation,
    this.duree,
    this.conditions,
    required this.status,
    required this.adminId,
    required this.createdAt,
    required this.updatedAt,
  });

  factory Offre.fromJson(Map<String, dynamic> json) {
    return Offre(
      id: json['id'] is String ? int.parse(json['id']) : json['id'],
      titre: json['titre'] ?? '',
      description: json['description'] ?? '',
      typeFormation: json['type_formation'],
      duree: json['duree'],
      conditions: json['conditions'],
      status: json['status'] ?? 'pending',
      adminId: json['admin_id'] is String ? int.parse(json['admin_id']) : (json['admin_id'] ?? 0),
      createdAt: DateTime.parse(json['created_at'] ?? DateTime.now().toIso8601String()),
      updatedAt: DateTime.parse(json['updated_at'] ?? DateTime.now().toIso8601String()),
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'titre': titre,
      'description': description,
      'type_formation': typeFormation,
      'duree': duree,
      'conditions': conditions,
      'status': status,
      'admin_id': adminId,
      'created_at': createdAt.toIso8601String(),
      'updated_at': updatedAt.toIso8601String(),
    };
  }

  bool get isValidated => status == 'validated';
  bool get isPending => status == 'pending';
  bool get isRejected => status == 'rejected';
}
